### PR TITLE
Give proper error while distributing a temp table

### LINKED ISSUE #4658: Distributing a temporary table
Distributing a temporary table gives an error message that is related to how temp tables are stored in postgres. It could be more informative to users if we gave an error like "Temporary tables cannot be distributed".

[local] talha@talha:9700-31535=# select create_reference_table('a');
ERROR:  unacceptable schema name "pg_temp_3"
DETAIL:  The prefix "pg_" is reserved for system schemas.
CONTEXT:  while executing command on localhost:9701
Time: 18.671 ms
[local] talha@talha:9700-31535=#

### PR DESCRIPTION
**Pull request by SaitTalhaNisanci**
DESCRIPTION: Improves error message while distributing a temporary table

Fixes #4658

### CODE REVIEW

File: src/backend/distributed/commands/create_distributed_table.c [hunk 1]
```diff
@@ -329,6 +330,9 @@ EnsureCitusTableCanBeCreated(Oid relationOid)
 	EnsureCoordinator();
 	EnsureRelationExists(relationOid);
 	EnsureTableOwner(relationOid);
+	Relation relation = relation_open(relationOid, AccessShareLock);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c:333]: To be more inline with the above, maybe move the code `EnsureTableNotTemporary` or such

File: src/test/regress/expected/isolation_drop_vs_all.out [hunk 2]
```diff
@@ -241,7 +241,7 @@ step s1-drop: DROP TABLE drop_hash;
 step s2-distribute-table: SELECT create_distributed_table('drop_hash', 'id'); <waiting ...>
 step s1-commit: COMMIT;
 step s2-distribute-table: <... completed>
-ERROR:  could not create distributed table: relation does not exist
+ERROR:  could not open relation with OID XXXX
```
Comment (onderkalaci) [inline src/test/regress/expected/isolation_drop_vs_all.out:244]: Why does this change? We call `EnsureRelationExists` before the new code?

**Review by onderkalaci — COMMENTED** [review]
Also, what about Citus local tables? Should we add a similar check to `ErrorIfUnsupportedCreateCitusLocalTable`?

**Comment by SaitTalhaNisanci** [comment]
>Also, what about Citus local tables? Should we add a similar check to ErrorIfUnsupportedCreateCitusLocalTable?

Makes sense but I am curious about why we don't call `EnsureCitusTableCanBeCreated` in `CitusLocalTable`, Othen than `EnsureRelationKindSupported`, the other checks should be valid for citus local tables as well I guess?

**Comment by SaitTalhaNisanci** [comment]
>Should we add a similar check to ErrorIfUnsupportedCreateCitusLocalTable?

We seem to be already erroring out for citus local tables:https://github.com/citusdata/citus/blob/master/src/test/regress/expected/citus_local_tables.out#L57

File: src/test/regress/expected/isolation_drop_vs_all.out (hunk 2 above)
Comment (SaitTalhaNisanci) [reply src/test/regress/expected/isolation_drop_vs_all.out:244]: I thin this error comes from a different line and we move the check earlier now, though I am not 100% sure why `EnsureRelationExists` doesn't error out before the below line:

```c
	EnsureCitusTableCanBeCreated(relationId);

	/* enable create_distributed_table on an empty node */
	InsertCoordinatorIfClusterEmpty();

	/*
	 * Lock target relation with an exclusive lock - there's no way to make
	 * sense of this table until we've committed, and we don't want multiple
	 * backends manipulating this relation.
	 */
	Relation relation = try_relation_open(relationId, ExclusiveLock);
	if (relation == NULL)
	{
		ereport(ERROR, (errmsg("could not create distributed table: "
							   "relation does not exist")));
	}

```

File: src/backend/distributed/commands/create_distributed_table.c [hunk 3]
```diff
@@ -1166,6 +1168,22 @@ EnsureRelationCanBeDistributed(Oid relationId, Var *distributionColumn,
...
 
 
+/*
+ * ErrorIfTemporaryTable errors out if the given table is a temporary table.
+ */
+static void
+ErrorIfTemporaryTable(Oid relationId)
+{
+	Relation relation = relation_open(relationId, AccessShareLock);
+	if (relation->rd_rel->relpersistence == RELPERSISTENCE_TEMP)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c:1178]: Minor: we can use `get_rel_persistence `

**Review by onderkalaci — APPROVED** [review]

File: src/backend/distributed/commands/create_distributed_table.c (hunk 3 above)
Comment (SaitTalhaNisanci) [reply src/backend/distributed/commands/create_distributed_table.c]: Thanks, it is definitely better to use that as it abstracts the details.

**Pull request merged**